sdk/data/azcosmos/cosmos_global_endpoint_manager.go

```diff
@@ -0,0 +1,186 @@
+	}
+
+	azResponse, err := gem.client.sendGetRequest(path, context.Background(), operationContext, nil, nil)
+	if err != nil {
+		return accountProperties{}, fmt.Errorf("failed to send GET request: %w", err)
+	}
+
+	properties, err := newAccountProperties(azResponse)
+	if err != nil {
+		return accountProperties{}, fmt.Errorf("failed to parse account properties: %w", err)
+	}
+
+	return properties, nil
+}
+
+// newAccountProperties parses the account properties from the HTTP response.
+func newAccountProperties(azResponse *http.Response) (accountProperties, error) {
+	properties := accountProperties{}
+	err := azruntime.UnmarshalAsJSON(azResponse, &properties)
+	if err != nil {
+		return properties, err
+	}
+	return properties, nil
+}
+
+// GetLocation returns the location for the given endpoint from the location cache.
+func (gem *globalEndpointManager) GetLocation(endpoint url.URL) string {
+	return gem.locationCache.getLocation(endpoint)
+}
+
+// MarkEndpointUnavailableForRead marks an endpoint as unavailable for read operations.
+func (gem *globalEndpointManager) MarkEndpointUnavailableForRead(endpoint url.URL) error {
+	return gem.locationCache.markEndpointUnavailableForRead(endpoint)
+}
+
+// MarkEndpointUnavailableForWrite marks an endpoint as unavailable for write operations.
+func (gem *globalEndpointManager) MarkEndpointUnavailableForWrite(endpoint url.URL) error {
+	return gem.locationCache.markEndpointUnavailableForWrite(endpoint)
+}
+
+// Update updates the location cache and the client's default endpoint based on the provided account properties.
+func (gem *globalEndpointManager) Update() error {
+	accountProperties, err := gem.GetAccountProperties()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve account properties: %w", err)
+	}
+
+	if err := gem.locationCache.update(accountProperties.WriteRegions, accountProperties.ReadRegions, gem.preferredRegions, &accountProperties.EnableMultipleWriteLocations); err != nil {
+		return fmt.Errorf("failed to update location cache: %w", err)
+	}
+	return nil
+}
+
+func (gem *globalEndpointManager) startBackgroundRefresh(refreshInterval time.Duration) {
+	gem.Lock()
+	defer gem.Unlock()
+
+	if gem.refreshTicker != nil {
+		// Background refresh already running, stop it first
+		gem.refreshTicker.Stop()
+		close(gem.refreshDone)
+	}
+
+	gem.refreshTicker = time.NewTicker(refreshInterval)
+	gem.refreshDone = make(chan struct{})
+	accountProperties, err := gem.GetAccountProperties()
+	if err != nil {
+		log.Write(azlog.EventResponse, fmt.Sprintf("Background refresh error: %v", err))
+		return
+	}
+	go func() {
+		for {
+			select {
+			case <-gem.refreshTicker.C:
+				err := gem.locationCache.update(accountProperties.WriteRegions, accountProperties.ReadRegions, gem.preferredRegions, &accountProperties.EnableMultipleWriteLocations)
+				if err != nil {
+					log.Write(azlog.EventResponse, fmt.Sprintf("Background refresh error: %v", err))
```

Comment:
Having a background goroutine silently swallow errors seems dubious.  How is a customer to know if an error occurred, and what is the recovery path?

Reply:
this is a siently error. it should not disrupt the connection rather us the lastest location cache.

Reply:
It sounds to me that it's not really an error then and should be removed.

Reply:
I'm still unclear on the value of this.

Reply:
In Cosmos SDKs (this is common to all languages), we have a background task that every 5 minutes refreshes the account information. The goal is to proactively detect any changes in account information (Available Regions, Consistency) that will affect how the client routes requests.

One common scenario is: https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/troubleshoot-sdk-availability#adding-a-region-to-an-account

If for whatever reason the refresh fails, it is not critical. Depending on the error, we might decide to do a retry (for example, for timeouts) or in some cases, we can even make it better and use other regional endpoints (there is a global DNS and there are regional endpoints with a different DNS each).